[Security] OAuth2 Introspection Endpoint (RFC7662)

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | none
| License       | MIT
| Doc PR        | symfony/symfony-docs#[**TODO**]

In addition to the excellent work of @vincentchalamon #48272, this PR allows getting the data from the OAuth2 Introspection Endpoint. This endpoint is defined in the [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662). It returns the following information that is used to retrieve the user:

* If the access token is active
* A set of claims that are similar to the OIDC one, including the `sub` or the `username`.

Example of configuration:

```yaml
framework:
    http_client:
        scoped_clients:
            oauth2.client:
                base_uri: 'https://authorization-server.example.com/introspection'
                scope: 'https://authorization-server\.example\.com'
                headers:
                    Authorization: 'Basic Y2xpZW50OnBhc3N3b3Jk' # Introspection Endpoint usually requires client authentication

security:
    firewalls:
        main:
            pattern: ^/
            access_token:
                token_handler:
                    oauth2: ~
                token_extractors: 'header' 
                realm: 'My API'
```

wouterj commented:
Thanks for yet another security feature proposal!
First a practical thing: We're entering feature freeze very shortly, meaning that we don't merge newly opened feature PRs for 6.3 anymore (unless really needed or very small). So we have some more time to polish this PR for 6.4 :)

I want to discuss a general remark that I don't think I've shared publicly yet. _(disclaimer: I now literally nothing about OAuth 2.0 and OIDC, except that it consists of a ton of different standards, so I can't determine if this applies to this PR)_

My vision for Symfony Security is to provide a minimal implementation following industry best practices based on standards and let community libraries/bundle add all "bonus features". We don't have the man power nor knowledge to maintain let's say all OAuth, SAML, Webauthn and MFA standards.

When adding new features, I think we must test it to those 3 conditions:
* Is the feature based on a standard? (which is clearly yes for this one)
* Is this feature required to have a working authentication/authorization system, within reason (i.e. if 90% of the apps use it when using OIDC)? (this is what I'm a bit unsure about for this specific PR)
* Is this feature following industry best practices? (e.g. this is why I pushed a bit to add a JWT implementation in the OIDC PR, given that the user info handler is generally not recommended for production)

cc @chalasr

chalasr commented:
I agree with Wouter.
Implementing all OAuth2/OIDC features in core is a non-goal for now. Better improve community packages that already supporting them or are meant to do so.

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+            // If the token is invalid or expired, the Introspection Endpoint will return 'active' => false
+            $claims = $this->client->request('POST', '',[
+                'headers' => [
+                    'Content-Type' => 'application/x-www-form-urlencoded',
```
not needed, that's already the default

Spomky replied:
👍🏽

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+        try {
+            // Call the Introspection Endpoint to retrieve the user info
+            // If the token is invalid or expired, the Introspection Endpoint will return 'active' => false
+            $claims = $this->client->request('POST', '',[
```
missing space before `[`

Spomky replied:
👌🏼

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+                ],
+            ])->toArray();
+
+            if (empty($claims['active'])) {
```
if (!($claims['active'] ?? false)) {

Spomky replied:
👌🏼

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+            if (empty($claims['active'])) {
+                throw new MissingClaimException(sprintf('"%s" claim not found on the authorization server response.', 'active'));
+            }
+            if ($claims['active'] !== true) {
```
`if ($claims['active']) {`

Spomky replied:
Fixed

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+            if ($claims['active'] !== true) {
+                throw new InactiveAccessTokenException('The access token is not active.');
+            }
+            if (empty($claims[$this->claim])) {
```
We try to avoid using `empty()` and using comparisons to boolean, unless a type check is required.
`if (!($claims[$this->claim] ?? false)) {`

Spomky replied:
👍🏽

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+            }
+
+            return new UserBadge($claims[$this->claim], fn () => $this->createUser($claims), $claims);
+        } catch (\Throwable $e) {
```
can we make this more specific somehow? catching any throwable is also something we try to avoid unless there are reasons

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/AccessToken/OAuth2/OAuth2TokenHandlerTest.php:
```diff
@@ -0,0 +1,74 @@
+        $this->assertEquals(new UserBadge($expected, fn () => $expectedUser, $claims), $userBadge);
+        $this->assertInstanceOf(OAuth2User::class, $actualUser);
+        $this->assertEquals($claims, $userBadge->getAttributes());
+        $this->assertEquals($claims['sub'], $actualUser->getUserIdentifier());
```
please use assertSame everywhere possible

Spomky replied:
Fixed

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/AccessToken/OAuth2/OAuth2TokenHandlerTest.php:
```diff
@@ -0,0 +1,74 @@
+        $this->assertEquals($claims['sub'], $actualUser->getUserIdentifier());
+    }
+
+    public function getClaims(): iterable
```
phpunit providers should be all static

Spomky replied:
👍🏽

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @see https://tools.ietf.org/html/rfc7662
+ *
+ * @experimental
```
we usually don't add experimental features to existing component: either a new component is introduced and it can be exprimental, or it's a new non-experimental feat of an existing component

Spomky replied:
👍🏽

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+
+/**
+ * The token handler validates the token on the authorization server.
+ * It uses the Introspection Endpoint as per the RFC7662
```
Missing blank line before: docblock should be formatted as one line for the title, a blank line, then longer description.

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/OAuth2Trait.php:
```diff
@@ -0,0 +1,53 @@
+            $claims[u($claim)->camel()->toString()] = $value;
+        }
+
+        if (isset($claims['updatedAt']) && '' !== $claims['updatedAt']) {
```
`if ('' !== ($claims['updatedAt'] ?? '')) {`

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/OAuth2Trait.php:
```diff
@@ -0,0 +1,53 @@
+            if ('' === $value || null === $value) {
+                continue;
+            }
+            $claims[u($claim)->camel()->toString()] = $value;
```
not sure this deserves the added dependency to the String component

Spomky replied:
The OIDC handler also uses it. If we can get rid of it, I will change that line. TBD with @vincentchalamon

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/OAuth2Trait.php:
```diff
@@ -0,0 +1,53 @@
+            $claims['updatedAt'] = (new \DateTimeImmutable())->setTimestamp($claims['updatedAt']);
+        }
+
+        if (\array_key_exists('emailVerified', $claims) && null !== $claims['emailVerified'] && '' !== $claims['emailVerified']) {
```
`if ('' !== ($claims['emailVerified'] ?? '')) {`, same below

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,76 @@
+ */
+final class Oauth2TokenHandler implements AccessTokenHandlerInterface
+{
+    use OAuth2Trait;
```
this trait is used only once, so let's inline?

nicolas-grekas commented on src/Symfony/Component/Security/Core/User/OAuth2User.php:
```diff
@@ -0,0 +1,132 @@
+        private ?string $sub = null,
+        private ?string $aud = null,
+        private ?string $iss = null,
+        private ?string $jti = null,
```
what about making all props public (readonly?) and removing all getters?

Spomky replied:
Correct. I changed the design. Looks good to me.

Spomky commented:
Hi @wouterj and @chalasr,

Many thanks for your feedback and advices. Hereafter my thoughts on this

> First a practical thing: We're entering feature freeze very shortly, meaning that we don't merge newly opened feature PRs for 6.3 anymore (unless really needed or very small). So we have some more time to polish this PR for 6.4 :)

No problem. I based it on the branch 6.3, but it does not matter to rebase to another branch.

> We don't have the man power nor knowledge to maintain let's say all OAuth, SAML, Webauthn and MFA standards.

Even if the community is large enough, I understand there is a risk to include specific authentication means that could be a problem in the future.

> When adding new features, I think we must always test it to those 3 conditions

I was not aware of those, but seem to be fair enough.

> Implementing all OAuth2/OIDC features in core is a non-goal for now. Better improve community packages that are already supporting them or are meant to do so.

Noted. Thanks. In this case, there is no need to go further here. I will continue that work in a side project. 👍

nicolas-grekas commented on src/Symfony/Component/Security/Core/User/OAuth2User.php:
```diff
@@ -0,0 +1,132 @@
+        ...$additionalClaims
+    ) {
+        if ((null === $sub || '' === $sub) && (null === $username || '' === $username)) {
+            throw new \InvalidArgumentException('The "sub" and "username" claims cannot be empty.');
```
`You must provide one of "sub" or "username" claims.`

Spomky replied:
OK. Done

nicolas-grekas commented on src/Symfony/Component/Security/Core/User/OAuth2User.php:
```diff
@@ -0,0 +1,132 @@
+        //    their own service-specific response names as top-level members
+        //    of this JSON object.
+        // ")
+        ...$additionalClaims
```
could this work as a regular array argument?

Spomky replied:
To be confirmed. As the object is called using the spread operator, this seems to be the safest way.
As it is inspired by the OIDC, the signature of the OIDCUser should also be changed.

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator_access_token.php:
```diff
@@ -61,5 +62,14 @@
+            abstract_arg('http client'),
+            service('logger')->nullOnInvalid(),
+            'sub',
+        ])
```
missing indentation after "set" line

Spomky replied:
Fixed

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OAuth2TokenHandlerFactory.php:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Configures a token handler for an Introspection Endpoint.
+ *
+ * @experimental
```
`@internal` instead

Spomky replied:
OK. Done

Spomky commented:
Hi,

I revised my mind and I would like to reopen this PR. I think this should be challenged a bit more.
Hereafter a (simplified) diagram showing where this new feature will be useful.

```mermaid
sequenceDiagram
    participant Client
    participant Authorization Server
    participant Resource Server
    Authorization Server->>Client: IDToken+Access Token (after consent+authentication)
    loop PR 48272 acts here
        Client->>Client: IDToken
        Client->>Authorization Server: Access Token (UserInfo Endpoint)
    end
    Client->>Resource Server: Access Token
    loop This PR acts here
        Resource Server->>Resource Server: Access Token (e.g. JWT, PASETO...)
        Resource Server->>Authorization Server: Access Token (Introspection Endpoint)
    end
```

As you can see, the features introduces with #48272 allow a Client application to authenticate end-users (by end-user, the OAuth spec means a person). The Client app can either use the IDToken (JWT) or the Access Token+UserInfo Endpoint to retreive the end-user details.
```mermaid
sequenceDiagram
    participant Client
    participant Authorization Server
    loop PR 48272 acts here
        Client->>Client: IDToken
        Client->>Authorization Server: Access Token (UserInfo Endpoint)
    end
```

This PR acts in a very similar way. The main differences are as follows:
* It is s not meant to be used by the Client, but the Resource Server
* The Resource Server is not supposed to receive the ID Token (unless Client + Resource Server = the same application)
* The Resource Server can directly use the Access Token if it can be verified (e.g. JWT)
* The "user" (consumer of the Resource Server) can be a end-user or an application (e.g. API to API communication)

```mermaid
sequenceDiagram
    participant Resource Server
    participant Authorization Server
    loop This PR acts here
        Resource Server->>Resource Server: Access Token (e.g. JWT, PASETO...)
        Resource Server->>Authorization Server: Access Token (Introspection Endpoint)
    end
```

The reason why I think this PR deserves more attention is that is cover the missing piece for covering most of the OAuth2 use cases.

I understand the point of view you say we could rely community on community works. And that's true that several solutions exist. But here, no third party library with exotic dependencies is needed<sup>1</sup>. We can directly use it with few configuration lines on e.g. ApiPlatform-based resource servers.

Also, I understand that Symfony does not aim at implementing all industry/web standards (even if all the Notifier channels make me think the opposite). However, OAuth2 is so widely used by the industry that it would be a shame not to support it, in particular after the works on the previous PRs. It is not a trendy technology, but a standard widely used for more than 10 years (and in continuous evolution).

At last but not least, the item Support for third-party authentication services (e.g. Auth0: https://auth0.com/) in [[RFC] Symfony Security rework tracking issue](https://github.com/symfony/symfony/issues/30914) proposed by @curry684 could be marked as checked as this feature will allow communication with [auth0/Okta](https://developer.okta.com/docs/reference/api/oidc/#introspect), [OneLogin](https://developers.onelogin.com/openid-connect/api/validate-token) and many more.

---

<sup>1</sup>: this does not mean community works are useless. They all have great features, but having too much external dependencies is something we all try to avoid.

vincentchalamon commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OAuth2TokenHandlerFactory.php:
```diff
@@ -0,0 +1,72 @@
+            ->arrayNode($this->getKey())
+                ->fixXmlConfig($this->getKey())
+                ->children()
+                    ->arrayNode('client')
```
[suggestion] Allow to create the client directly here, like in [OidcUserInfoTokenHandlerFactory](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php). It would also add consistency with `oidc_user_info` usage.

Spomky replied:
HI @vincentchalamon,

I am not sure to clearly understand. The behavior looks identical. What have I missed?

vincentchalamon replied:
It's fixed, maybe I read it wrong or you already fixed it. Thanks :-)

Spomky replied:
👍🏽 Thanks!

OskarStark commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OAuth2TokenHandlerFactory.php:
```diff
@@ -0,0 +1,72 @@
+use Symfony\Component\HttpClient\HttpClient;
+
+/**
+ * Configures a token handler for an OIDC server.
```
Is this comment correct?

Spomky replied:
Indeed. I updated this comment.

OskarStark commented on src/Symfony/Component/Security/Http/Tests/AccessToken/OAuth2/OAuth2TokenHandlerTest.php:
```diff
@@ -0,0 +1,71 @@
+
+    public static function getClaims(): iterable
+    {
+        yield ['Z5O3upPC88QrAjx00dis'];
```
Why a data provider just for one?

Spomky replied:
The first iteration of this PR had several use cases and I simplified it to that unique one.
But yes it is not relevant now and I removed it.

Spomky commented:
Rebased for targeting Symfony 7.1

chalasr commented on src/Symfony/Bundle/SecurityBundle/CHANGELOG.md:
```diff
@@ -24,12 +24,14 @@ CHANGELOG
  * Deprecate the `require_previous_session` config option. Setting it has no effect anymore
  * Add `LogoutRouteLoader`
 
+
```
should be reverted

chalasr commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Exception/MissingClaimException.php:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @experimental
+ */
+class MissingClaimException extends AuthenticationException
```
I'm not sure we need this very specific exception given it's immediately caught where it's throw and wrapped in a `BadCredentialsException`. I suggest to throw `BadCrendentialsException`s with precise messages where needed instead (see https://github.com/symfony/symfony/pull/50695 - I'll finish it at some point so better not continue this way)

chalasr reviewed (approved):
I changed my mind, given we have OAuth/OIDC support let's do it right + as complete as possible. 👍 once the small comments addressed & PR rebased

Spomky commented:
That's good news!
The PR is now rebased and I addressed the remarks raised earlier.

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OAuth2TokenHandlerFactory.php:
```diff
@@ -0,0 +1,72 @@
+                          ->isRequired()
+                        ->beforeNormalization()
+                            ->ifString()
+                           ->then(static function ($v): array { return ['id' => $v]; })
```
```suggestion
                            ->then(static fn ($v) => ['id' => $v])
```

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator_access_token.php:
```diff
@@ -135,5 +136,13 @@
+        ->args([
+            abstract_arg('http client'),
+            service('logger')->nullOnInvalid(),
+        ])
```
indentation is off by one

Spomky replied:
Indentation should be fine now.

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OAuth2TokenHandlerFactory.php:
```diff
@@ -0,0 +1,72 @@
+                    ->addError('You cannot use the "oauth2" token handler since the HttpClient component is not installed. Try running "composer require symfony/http-client".');
+            } else {
+                $container->register($clientDefinitionId, HttpClient::class)
+                    ->setFactory([HttpClient::class, 'create'])
```
we should use the existing http_client service instead, so that we get integration with the profiler for free
or is there a reason why this is a separate service?

Spomky replied:
Indeed. This is now gone.

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OAuth2TokenHandlerFactory.php:
```diff
@@ -0,0 +1,72 @@
+                $container->register($clientDefinitionId, 'stdClass')
+                    ->addError('You cannot use the "oauth2" token handler since the HttpClient component is not installed. Try running "composer require symfony/http-client".');
+            } else {
+                $container->register($clientDefinitionId, HttpClient::class)
```
if we need that service, it should be defined in a service-declaration file

Spomky replied:
No need of it after all.

nicolas-grekas commented on src/Symfony/Component/Security/Core/User/OAuth2User.php:
```diff
@@ -0,0 +1,72 @@
+/**
+ * UserInterface implementation used by the access-token security workflow with an OIDC server.
+ *
+ * @experimental
```
not sure this is useful, the API is standard

Spomky replied:
Indeed. Gone.

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/OAuth2/Oauth2TokenHandler.php:
```diff
@@ -0,0 +1,97 @@
+                ],
+            ])->toArray();
+
+            if (empty($claims['sub']) && empty($claims['username'])) {
```
this and adjust below to use the new vars:
```suggestion
            $sub = \is_array($claims['sub']) ? $claims['sub'] : null;
            $username = \is_array($claims['username']) ? $claims['username'] : null;
            if (!$sub && !$username)) {
```

Spomky replied:
Thank you for the review.
This part is now up to date.

nicolas-grekas commented on src/Symfony/Component/Security/Http/AccessToken/Oidc/OidcTokenHandler.php:
```diff
@@ -85,19 +84,16 @@ public function getUserBadgeFrom(string $accessToken): UserBadge
-            $claimCheckerManager->check($claims);
+            ]))->check($claims);
 
             if (empty($claims[$this->claim])) {
```
empty is dangerous and we try to avoid using it

Spomky replied:
Should be fine in the last revision of this PR.
Thanks.

nicolas-grekas reviewed (commented):
Here are some CS suggestions, please rebase also.

welcoMattic commented on src/Symfony/Component/Security/Http/Tests/AccessToken/OAuth2/OAuth2TokenHandlerTest.php:
```diff
@@ -0,0 +1,63 @@
+        ];
+        $expectedUser = new OAuth2User(...$claims);
+
+        $responseMock = $this->createMock(ResponseInterface::class);
+        $responseMock->expects($this->once())
+            ->method('toArray')
+            ->willReturn($claims);
+
+        $clientMock = $this->createMock(HttpClientInterface::class);
+        $clientMock->expects($this->once())
+            ->method('request')->with('POST', '', [
+                'body' => [
+                    'token' => $accessToken,
+                    'token_type_hint' => 'access_token',
+                ],
+            ])
+            ->willReturn($responseMock);
```
Why not using `MockHttpClient` here? https://symfony.com/doc/current/http_client.html#http-client-and-responses

Spomky replied:
Indeed. MockHttpClient is now used in the last revision of this PR.
Thank you.

Spomky committed:
[Security] OAuth2 Introspection Endpoint (RFC7662)

In addition to the excellent work of @vincentchalamon #48272, this PR allows getting the data from the OAuth2 Introspection Endpoint. This endpoint is defined in the [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662). It returns the following information that is used to retrieve the user:

* If the access token is active
* A set of claims that are similar to the OIDC one, including the `sub` or the `username`.

fabpot commented:
Thank you @Spomky.
